### PR TITLE
feat: MoneymanDash storage adapter

### LINKED ISSUE #785: Feature: Add MoneymanDash storage adapter
Add MoneymanDash storage adapter.

### PR DESCRIPTION
**Pull request by daniel-hauser**
Adds storage adapter for moneyman-dash. Supports mm_ token format, sends transactions to /ingest, logs to /logs as plain text with X-Run-Id header. 17 tests. Closes #785

## Summary by CodeRabbit

* **New Features**
  * Added a storage backend to upload completed transactions and send logs to an external ingestion endpoint.
  * Introduced per-run context tracking with unique run IDs to correlate saves and logs.
  * Added configurable schema support and an optional log-upload capability on storage implementations.
  * Enhanced log transmission to forward logs to configured storage backends in addition to existing channels.

* **Tests**
  * Added a comprehensive test suite covering token parsing, payload validation, transaction filtering, error handling, progress callbacks, and log submission.

### CODE REVIEW

File: src/bot/storage/moneyman.test.ts [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+      // Call without runContext
+      await storage.saveTransactions([transactionRow({})], async () => {});
+
+      const callArg = fetchMock.mock.calls[0][1];
+      const body = JSON.parse(callArg.body);
+
+      // Should not include runId if not in context
+      expect(body.runId).toBeUndefined();
+    });
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.test.ts:354]: [copilot on behalf of Daniel] Good catch — fixed. The assertion now correctly checks \ody.metadata.runId\ instead of \ody.runId\.

File: src/bot/storage/moneyman.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+      const base64 = encoded
+        .replace(/-/g, "+")
+        .replace(/_/g, "/")
+        .padEnd(encoded.length + ((4 - (encoded.length % 4)) % 4), "=");
+      const decoded = JSON.parse(
+        Buffer.from(base64, "base64").toString("utf-8"),
+      );
+      this.endpoint = decoded.u; // full ingest URL, e.g. https://...convex.site/ingest
+      this.token = decoded.k; // bearer secret
+      return;
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts]: [copilot on behalf of Daniel] Added validation — \parseToken()\ now throws if \decoded.u\ or \decoded.k\ are missing. The constructor catches this (by design — other storage backends may be configured), but the error is logged clearly. Added a test to verify the graceful handling.

File: src/bot/storage/moneyman.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+    if (!runId) {
+      logger("Warning: No runId available, skipping log upload");
+      return;
+    }
+
+    logger("Sending logs");
+
+    // Derive /logs URL from /ingest URL
+    const logsUrl = this.endpoint.replace(/\/ingest$/, "/logs");
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts]: [copilot on behalf of Daniel] Added the guard — if the endpoint doesn't end with \/ingest\, we log a warning and skip the log upload. Added a test to verify this behavior.

File: src/bot/storage/moneyman.test.ts [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+      // Now fail the logs upload
+      fetchMock.mockResolvedValue({
+        ok: false,
+        status: 500,
+        statusText: "Internal Server Error",
+      });
+
+      // Should not throw, just log the error
+      await storage.sendLogs("logs");
+    });
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.test.ts:519]: [copilot on behalf of Daniel] Fixed — test now uses \/ingest\ endpoint so the fetch mock is properly exercised and the 500-error handling path is tested.

File: src/bot/storage/moneyman.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+  constructor(private config: MoneymanConfig) {
+    if (this.canSave()) {
+      try {
+        this.parseToken();
+      } catch (e) {
+        logger("Failed to parse token", e);
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts:52]: [copilot on behalf of Daniel] Fixed — \canSave()\ now checks \	his.endpoint && this.token\ (parsed results) instead of config existence. Constructor uses config check directly to decide whether to attempt parsing. Invalid tokens gracefully exclude the storage from the active list.

File: src/bot/storage/moneyman.ts [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    // endpoint is the full ingest URL from the mm_ token
+    const response = await fetch(this.endpoint, {
+      method: "POST",
+      headers: {
+        Authorization: `Bearer ${this.token}`,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(payload),
+    });
+
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts]: [copilot on behalf of Daniel] Good point — updated to use \Promise.all\ for consistency with other storage adapters like WebPostStorage.

File: src/bot/storage/moneyman.ts [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+  private endpoint?: string;
+  private token?: string;
+  private lastRunId?: string;
+
+  constructor(private config: MoneymanConfig) {
+    if (this.canSave()) {
+      try {
+        this.parseToken();
+      } catch (e) {
+        logger("Failed to parse token", e);
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts:49]: [copilot on behalf of Daniel] Fixed — \canSave()\ now returns \Boolean(this.endpoint && this.token)\, so parse failures gracefully exclude this storage. The constructor still catches the error (by design — other storage backends may be configured), but the storage won't be in the active list.

File: src/bot/storage/moneyman.ts [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+  private buildIngestionPayload(txns: TransactionRow[]): IngestionPayload {
+    const uniqueAccounts = new Set(txns.map((t) => t.account));
+
+    return {
+      metadata: {
+        scrapedAt: new Date().toISOString(),
+        scrapedBy: "moneyman",
+        accounts: uniqueAccounts.size,
+        added: txns.length,
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts:235]: [copilot on behalf of Daniel] Added a comment explaining why these are 0 — the \	xns\ array is already filtered to non-pending transactions by the caller, and moneyman doesn't track skipped/highlighted at this level.

File: src/bot/storage/moneyman.test.ts [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+      const completedTx = transactionRow({
+        status: TransactionStatuses.Completed,
+      });
+      const pendingTx = transactionRow({
+        status: TransactionStatuses.Pending,
+      });
+
+      fetchMock.mockResolvedValue(
+        mockSuccessResponse({ added: 1, pending: 1 }),
+      );
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.test.ts:243]: [copilot on behalf of Daniel] Fixed — mock overrides now use \	ransactionsAdded\ to match the real \IngestionResponse\ contract.

File: src/bot/storage/moneyman.test.ts [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+      const token = makeToken("https://api.example.com", "secret123");
+
+      const txns = [
+        transactionRow({ account: "1234" }),
+        transactionRow({ account: "1234" }), // duplicate
+        transactionRow({ account: "5678" }),
+        transactionRow({ account: "9012" }),
+      ];
+
+      fetchMock.mockResolvedValue(mockSuccessResponse({ added: 4 }));
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.test.ts]: [copilot on behalf of Daniel] Fixed — changed to \	ransactionsAdded: 4\ to match the API contract.

File: src/utils/secure-log.ts [hunk 11]
```diff
@@ -38,10 +40,28 @@ export async function sendAndDeleteLogFile() {
...
     }
+
+    // Send logs to moneyman storage if configured
+    const moneymanStorage = storages.find(
+      (s) => s instanceof MoneymanDashStorage,
+    ) as MoneymanDashStorage | undefined;
+
+    if (moneymanStorage) {
+      logToPublicLog(`Sending logs to moneyman storage`, logger);
+      try {
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/utils/secure-log.ts]: [copilot on behalf of Daniel] Good catch — refactored \secure-log.ts\ to be generic. It now iterates all storages with \sendLogs\ (no \instanceof\ check), and the warning accounts for all log destinations (Telegram + any storage with \sendLogs\).

File: src/utils/secure-log.ts [hunk 12]
```diff
@@ -38,10 +40,28 @@ export async function sendAndDeleteLogFile() {
   }
 
   try {
+    // Read log file content for storage uploads
+    const logContent = readFileSync(logFilePath, "utf-8");
+
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/utils/secure-log.ts]: [copilot on behalf of Daniel] Fixed — \
eadFileSync\ is now deferred until we confirm there are storages with \sendLogs\. Telegram upload still streams from path.

File: src/bot/storage/moneyman.ts [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+        `Failed to post transactions: ${response.status} ${response.statusText}`,
+      );
+    }
+
+    const result = (await response.json()) as IngestionResponse;
+
+    logger(`Ingestion created: ${result.ingestionId}`);
+
+    const stats = createSaveStats("MoneymanDashStorage", "moneyman", txns);
+    stats.added = result.transactionsAdded;
```
[the remark this answers is not in the document]
Comment (daniel-hauser) [reply src/bot/storage/moneyman.ts:174]: [copilot on behalf of Daniel] Good catch — added runtime validation of the response shape. Now throws if `transactionsAdded` is not a number or `ingestionId` is not a string. Added a test for unexpected response shape as well.

**Comment by daniel-hauser** [comment]
[copilot on behalf of Daniel]

`IngestionMetadata` in `moneyman.ts` re-declares the `accountStatuses` shape inline instead of importing `AccountStatus` from `../../types.js`. Since `AccountStatus` is already exported there, the interface should use it:

```typescript
import type { AccountStatus } from "../../types.js";

interface IngestionMetadata {
  scrapedAt: string;
  scrapedBy: string;
  accounts: number;
  added: number;
  pending: number;
  skipped: number;
  highlightedTransactions: number;
  runId?: string;
  accountStatuses?: AccountStatus[];
}
```

Avoids duplicate type definitions that can drift out of sync.

**Comment by daniel-hauser** [comment]
[copilot on behalf of Daniel]

## PR Description Verification

The description says:
> "Adds storage adapter for moneyman-dash. Supports mm_ token format, sends transactions to /ingest, logs to /logs as plain text with X-Run-Id header. 17 tests. Closes #785"

**Accuracy check:**
- ✅ Storage adapter: `MoneymanDashStorage` class added
- ✅ mm_ token format: `mm_<base64url({ u, k })>` with URL validation, protocol check, prototype pollution protection
- ✅ Sends transactions to endpoint URL from token
- ✅ Logs to `{endpoint}/logs` as `text/plain` with `X-Run-Id` header
- ⚠️ Test count: File has ~17-20 test cases (hard to count exactly from diff, but ballpark is right)

**Missing from description** (significant cross-cutting changes not mentioned):
1. **`TransactionStorage` interface changed** — new `context?: SaveContext` param on `saveTransactions()` + new optional `sendLogs?(logs: string)` method. This is a breaking interface change affecting all storage implementations.
2. **`RunContext` via AsyncLocalStorage** — `bot/index.ts` wraps entire scraper run in `runContextStore.run({ runId })` for correlation
3. **`SaveContext` / `AccountStatus` types** — per-account scraping status (success/failure/errorType) passed through to storage
4. **`secure-log.ts` rewrite** — logs now forwarded to ALL storages implementing `sendLogs()`, not just Telegram
5. **`config.schema.ts`** — `MoneymanDashSchema` Zod validation added
6. **`storage/index.ts`** — builds `SaveContext` from `AccountScrapeResult[]` and passes to all storages

Suggested updated description:
> Adds MoneymanDash storage adapter with mm_ token format. Sends transactions to /ingest, logs to /logs (text/plain, X-Run-Id). Extends TransactionStorage interface with SaveContext and optional sendLogs(). Adds RunContext (AsyncLocalStorage) for per-run correlation. Updates secure-log.ts to forward logs to all storage backends. 17 tests. Closes #785

## Code Review Findings

### Issues to consider

1. **No fetch timeout** — Both `saveTransactions()` and `sendLogs()` use bare `fetch()` without `AbortController` timeout. In a scheduled scraper, a hung connection could block the entire process indefinitely.
   ```ts
   // Suggestion: add timeout
   const controller = new AbortController();
   const timeout = setTimeout(() => controller.abort(), 30_000);
   const response = await fetch(endpoint, { signal: controller.signal, ... });
   clearTimeout(timeout);
   ```

2. **`readFileSync` for log upload** — `secure-log.ts` reads the entire log file into memory with `readFileSync(logFilePath, "utf-8")`. For very large log files this could cause memory pressure. Consider streaming or chunking for large files (low priority — scraper logs are typically small).

3. **Sequential log sending** — `secure-log.ts` sends logs to storages one by one in a `for` loop. Could use `Promise.allSettled()` for parallelism (minor — typically only one storage has `sendLogs`).

### Code quality (positive)
- ✅ Prototype pollution protection on token parsing
- ✅ URL protocol validation (http/https only)
- ✅ Graceful error handling in constructor (catches parseToken failures)
- ✅ Defensive `lastRunId` fallback chain
- ✅ Good test coverage: token parsing edge cases, pending filter, error paths, account statuses
- ✅ Clean separation of concerns (buildIngestionPayload as private method)
